### DISCUSSION TITLE
Self hosted spin on nomad cluster how to achieve 0 downtime deployments

### DISCUSSION THREAD

**Discussion by Jamahl-Carter** [spinframework/spin#2549] [General]
I'm using `spin deploy` to push app changes to my self hosted nomad cluster and am consistently observing 404 responses for a brief window after the release.

I believe this is while the traffic is being re-routed to consume the new app version, is it possible to configure this to use the old version until the new version is ready ? Or any other method to remove this window of downtime ?

Thanks in advance for your help !

**Comment by lann** [comment]
Spin doesn't have a mechanism to do this today. You would typically achieve this by combining a load balancer with rolling restarts (or similar) so that both the old and new versions of the app are (briefly) being served at the same time.

**Reply by Jamahl-Carter** [reply]
Thanks for the speedy reply ! I'm not sure I fully understand, say I had 5 spin instances and apply a rolling restart:
- For the changed handlers, how would I ensure traffic is only routed when its ready (and avoid the 404 etc..) ?
- Is there some kind of health/readiness check I can configure for to inform the rolling restart on whether the instance is ready ?

If this release flow can't work then I'll likely opt for a canary style approach where I manually approve changes before they go live (or perhaps add a static timer if I'm feeling hacky).

Lastly, is there any planned works to better support/document best practices for this flow ? Zero downtime releases seems like an area which would be of particular interest for people looking to adopt this technology in production. Thanks again for your insights.

**Reply by lann** [reply]
Spin itself responds on `/.well-known/spin/health` but doesn't reflect anything about your actual app so you might want your own health endpoint.

**Reply by lann** [reply]
Nomad supports a number of different zero downtime strategies but the details will depend on your specific infrastructure.

There is also https://www.spinkube.dev/ which handles many of these things in a kubernetes environment.

**Reply by Jamahl-Carter** [reply]
Thanks for the info, I think k8 runs a little heavy for my use case but I'll look through their docs to see if I can glean some insights.

In the case where I have a "working" http spin trigger with 0 logic issues (lets say it just returns "hello world"), will I be able to use `/.well-known/spin/health`  to inform when my trigger is rout-able  (i.e. when I attempt to hit it, it will invoke my handler logic and not 404 etc..) ?

**Reply by lann** [reply]
> will I be able to use /.well-known/spin/health to inform when my trigger is rout-able (i.e. when I attempt to hit it, it will invoke my handler logic and not 404 etc..) ?

Yep. It's really more of a "readiness" check than a health check.

**Reply by Jamahl-Carter** [reply]
Perfect - this mechanism should work for my use case. Thanks Iann, I'll resolve this thread.
